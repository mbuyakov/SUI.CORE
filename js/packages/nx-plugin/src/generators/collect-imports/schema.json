{
  "$schema": "http://json-schema.org/schema",
  "$id": "CollectImports",
  "title": "",
  "type": "object",
  "properties": {
  },
  "required": []
}
