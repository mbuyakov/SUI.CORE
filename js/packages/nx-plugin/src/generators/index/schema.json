{
  "$schema": "http://json-schema.org/schema",
  "$id": "Index",
  "title": "",
  "type": "object",
  "properties": {}
}
