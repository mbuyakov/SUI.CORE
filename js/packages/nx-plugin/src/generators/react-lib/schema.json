{
  "$schema": "http://json-schema.org/schema",
  "$id": "Lib",
  "title": "",
  "type": "object",
  "properties": {
    "type": {
      "type": "string",
      "enum": ["lib","ui","util","deps", "host", "bundle"],
      "x-prompt": "What type would you like to use?"
    },
    "name": {
      "type": "string",
      "x-prompt": "What name would you like to use?"
    }
  },
  "required": ["type", "name"]
}
