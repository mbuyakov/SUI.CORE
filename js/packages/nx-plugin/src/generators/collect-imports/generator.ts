import {readJson, Tree} from "@nx/devkit";
import {CollectImportsGeneratorSchema} from "./schema";
import {visitAllFiles, visitAllProjects} from "../../utils/visitors";
import {oldPackages} from "../../utils/oldPackages";
import {astQuery} from "@sui/lib-typescript-ast";
import {
  isClassDeclaration,
  isExportAssignment,
  isExportDeclaration,
  isFunctionDeclaration,
  isInterfaceDeclaration,
  isTypeAliasDeclaration,
  isVariableStatement,
  SyntaxKind
} from "typescript";
import {logWithPrefix, setSpinnerText} from "../../utils/logger";
import {importsByNewPackage} from "../sui-migrator/modules/support/remapOldPackages";
import {parseExportDeclaration} from "./modules/parseExportDeclaration";
import {parseVariableStatement} from "./modules/parseVariableStatement";
import {parseCommonNamedNode} from "./modules/parseCommonNamedNode";

const naturalSorter = (a, b) => a.toLowerCase().localeCompare(b.toLowerCase());
const MAP_FILE_PATH = "packages/nx-plugin/src/generators/sui-migrator/modules/support/remapOldPackages.ts";
const newImportsByNewPackage: { [index: string]: string[] } = {};

function addExport(packageName: string, name: string) {
  if (!importsByNewPackage[packageName]?.includes(name)) {
    logWithPrefix("collectImport", `Add export ${name} from ${packageName}`);
  }
  newImportsByNewPackage[packageName] = newImportsByNewPackage[packageName] || [];
  newImportsByNewPackage[packageName].push(name);
}

export function collectImportsGenerator(tree: Tree, options: CollectImportsGeneratorSchema) {
  const content = tree.read(MAP_FILE_PATH).toString();
  visitAllProjects(tree, project => {
    const packageName = readJson(tree, `${project.root}/package.json`).name;

    if (project.name === "nx-plugin"
      // Don't parse old modules
      || oldPackages.includes(packageName)
    ) {
      return;
    }

    visitAllFiles(tree, project.sourceRoot, filePath => {
      if (
        (
          !filePath.endsWith(".ts")
          && !filePath.endsWith(".tsx")
        )
        || filePath.includes(".stories.")
      ) {
        return;
      }

      const content = tree.read(filePath).toString();
      const imports: string[] = [];

      if (!content.includes("export") || content.includes("SKIP_IMPORT_COLLECTING")) {
        return;
      }

      astQuery(content, "ExportKeyword").forEach(it => {
        it = it.parent;

        if (isExportAssignment(it)) {
          return;
        } else if (isExportDeclaration(it)) {
          imports.push(...parseExportDeclaration(it));
        } else if (isVariableStatement(it)) {
          imports.push(...parseVariableStatement(it));
        } else if (isFunctionDeclaration(it) || isTypeAliasDeclaration(it) || isClassDeclaration(it) || isInterfaceDeclaration(it)) {
          imports.push(parseCommonNamedNode(it));
        } else {
          throw new Error(`unknown parent type ${SyntaxKind[it.kind]} from ${filePath}`);
        }
      });
      imports.forEach(it => addExport(packageName, it));
    });

    setSpinnerText("common");
    importsByNewPackage[packageName]?.forEach(it => {
      if (!newImportsByNewPackage[packageName]?.includes(it)) {
        // language=text format=false
        logWithPrefix("collectImport", `Delete export ${it} from ${packageName}`);
      }
    });
  });

  Object.keys(newImportsByNewPackage).forEach(it => {
    newImportsByNewPackage[it].sort(naturalSorter);
  });

  const newContent = `
///////////////////////////////////
//FILE IS AUTOGENERATED. DON'T EDIT
///////////////////////////////////
export const importsByNewPackage = ${JSON.stringify(newImportsByNewPackage, Object.keys(newImportsByNewPackage).sort(), 2)};
  `.trim() + "\n";


  if (content != newContent) {
    tree.write(MAP_FILE_PATH, newContent);
  }
}

export default collectImportsGenerator;
