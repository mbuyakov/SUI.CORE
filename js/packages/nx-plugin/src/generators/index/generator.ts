import {Tree} from "@nx/devkit";
import {IndexGeneratorSchema} from "./schema";
import {visitAllFolders, visitAllProjects} from "../utils";
import {logSymbols, logWithPrefix} from "../../utils/logger";

const naturalSorter = (a, b) => a.toLowerCase().localeCompare(b.toLowerCase());

export async function indexGenerator(tree: Tree, options: IndexGeneratorSchema) {
  visitAllProjects(tree, project => {
    if (project.name === "nx-plugin") {
      return;
    }
    visitAllFolders(tree, project.sourceRoot, folderPath => {
      const filePath = `${folderPath}/index.ts`;
      const content = tree.exists(filePath) ? tree.read(filePath).toString() : "";
      let newContent = "";

      const isCustom = content
        .split("\n")
        .some(row => !row.startsWith("export * from") && row.length);

      if (isCustom) {
        logWithPrefix("indexGenerator", "index.ts has something other than \"export * from ...\". Ignore", logSymbols.warning);
        return;
      }

      const allContent = tree.children(folderPath);

      const folders = allContent
        .filter(it => !tree.isFile(`${folderPath}/${it}`))
        .sort(naturalSorter);

      const files = allContent
        .filter(it =>
          tree.isFile(`${folderPath}/${it}`)
          && it !== "index.ts"
          && !it.includes(".spec.")
          && !it.includes(".test.")
          && !it.includes(".stories.")
          && !it.endsWith(".less")
        )
        .sort(naturalSorter);
      if (folders.length) {
        folders.forEach(it => newContent += `export * from "./${it}";\n`);
      }

      if (files.length) {
        if (folders.length) {
          newContent += "\n";
        }
        files.forEach(it => newContent += `export * from "./${it.split(".")[0]}";\n`);
      }

      if (newContent !== content) {
        tree.write(filePath, newContent);
      }
    });
  });
}

export default indexGenerator;
