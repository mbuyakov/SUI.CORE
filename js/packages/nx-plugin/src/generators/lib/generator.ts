import {Tree} from "@nx/devkit";
import {LibGeneratorSchema} from "./schema";
import * as jsGen from "@nx/js/src/generators/library/library";
import {commonTweaks} from "../../utils/commonTweaks";

export async function libGenerator(tree: Tree, options: LibGeneratorSchema) {
  const name = `${options.type}-${options.name}`;

  await jsGen.libraryGenerator(tree, {
    name,
    bundler: "tsc",
    linter: "eslint",
    strict: true,
    testEnvironment: "jsdom",
    unitTestRunner: "jest",
    importPath: `@sui/${name}`
  });

  commonTweaks(tree, options.type, name);
}

export default libGenerator;
