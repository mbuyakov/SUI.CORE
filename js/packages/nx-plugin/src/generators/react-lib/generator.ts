import {Tree} from "@nx/devkit";
import {ReactLibGeneratorSchema} from "./schema";
import * as reactGen from "@nx/react";
import {Linter} from "@nx/linter";
import {commonTweaks} from "../utils";

export async function reactLibGenerator(tree: Tree, options: ReactLibGeneratorSchema) {
  const name = `${options.type}-${options.name}`;

  await reactGen.libraryGenerator(tree, {
    name,
    linter: Linter.EsLint,
    component: true,
    compiler: "babel",
    style: "@emotion/styled",
    unitTestRunner: "jest",
    bundler: "rollup",
    publishable: true,
    importPath: `@sui/${name}`,
    strict: true
  });

  commonTweaks(tree, options.type, name);
}


export default reactLibGenerator;
