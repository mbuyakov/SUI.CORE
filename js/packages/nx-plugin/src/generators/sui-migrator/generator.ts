import {Tree} from "@nx/devkit";
import {getProjects} from "nx/src/generators/utils/project-configuration";
import {SuiMigratorGeneratorSchema} from "./schema";
import {remapImports} from "./remapImports";
import {remapIcons} from "./remapIcons";
import {chalk} from "./util";
import {dedupeImport} from "./dedupeImport";

function visitAllFiles(tree: Tree, path: string, callback: (filePath: string) => void) {
  tree.children(path).forEach((fileName) => {
    const filePath = `${path}/${fileName}`;
    console.log(chalk.grey(`Process path ${filePath}`));
    if (!tree.isFile(filePath)) {
      visitAllFiles(tree, filePath, callback);
    } else {
      callback(filePath);
    }
  });
}

export async function suiMigratorGenerator(tree: Tree, schema: SuiMigratorGeneratorSchema) {
  const projects = getProjects(tree);
  projects.forEach(project => {
    console.log(`Process project ${project.name}`);
    visitAllFiles(tree, project.sourceRoot, (filePath) => {
      const fileEntry = tree.read(filePath);
      const content = fileEntry.toString();
      let newContent = content.toString();

      newContent = remapImports(project.name, newContent);
      newContent = remapIcons(newContent);
      newContent = dedupeImport(newContent);

      if (newContent !== content) {
        tree.write(filePath, newContent);
      }
    });
  });
}

export default suiMigratorGenerator;
