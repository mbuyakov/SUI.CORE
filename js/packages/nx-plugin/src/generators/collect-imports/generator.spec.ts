import {createTreeWithEmptyWorkspace} from "@nx/devkit/testing";
import {ProjectConfiguration, readProjectConfiguration, Tree} from "@nx/devkit";

import {collectImportsGenerator, MAP_FILE_PATH} from "./generator";
import {CollectImportsGeneratorSchema} from "./schema";
import libGenerator from "../lib/generator";

describe("collect imports", () => {
  let tree: Tree;
  let config: ProjectConfiguration;
  const options: CollectImportsGeneratorSchema = {};

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace();
    await libGenerator(tree, {
      type: "lib",
      name: "test"
    });
    config = readProjectConfiguration(tree, "lib-test");
  });

  it("should collect imports for test lib", async () => {
    tree.write(MAP_FILE_PATH, "");
    await collectImportsGenerator(tree, options);
    expect(tree.read(MAP_FILE_PATH).toString().trim()).toMatchSnapshot();
  });
});
