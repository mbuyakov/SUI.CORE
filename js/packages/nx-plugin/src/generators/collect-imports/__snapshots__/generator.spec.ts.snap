// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collect imports should collect imports for test lib 1`] = `
"///////////////////////////////////
//FILE IS AUTOGENERATED. DON'T EDIT
///////////////////////////////////
export const importsByNewPackage = {
  \\"@sui/lib-test\\": [
    \\"libTest\\"
  ]
};"
`;
